File: rust/src/game/systems/combat.rs
use crate::game::components::combat::{ AttackType, Health, HealthChanged, Stats, resolve_damage };
use crate::game::systems::rng::GameRng;
use bevy::prelude::*;

/// Event describing a single attack intent. Minimal and cheap.
#[derive(Event)]
pub struct AttackEvent {
    pub attacker: Entity,
    pub target: Entity,
    pub weapon: AttackType, // small enum (passed by ref to resolve)
    pub ability_multiplier_bp: Option<u16>, // 10000 = 1.0
}

/// Consume AttackEvent and apply damage.
/// Note: `events` must be provided and the event type registered with `app.add_event::<AttackEvent>()`.
pub fn apply_attack_events(
    mut ev_reader: EventReader<AttackEvent>,
    mut health_q: Query<&mut Health>,
    stats_q: Query<&Stats>,
    mut rng: ResMut<GameRng>,
    mut health_changed: EventWriter<HealthChanged> // emit health changes for other systems
) {
    for ev in ev_reader.read() {
        // get attacker stats
        let attacker_stats = match stats_q.get(ev.attacker) {
            Ok(s) => s,
            Err(_) => {
                continue;
            } // attacker missing stats to skip
        };

        // get target stats
        let target_stats = match stats_q.get(ev.target) {
            Ok(s) => s,
            Err(_) => {
                continue;
            } // target missing stats -> skip or treat as zero-def (choice)
        };

        // apply damage if target has health
        if let Ok(mut health) = health_q.get_mut(ev.target) {
            let roll = rng.next_u16_in_basis();
            let (damage, was_crit) = resolve_damage(
                &mut *health,
                attacker_stats,
                target_stats,
                &ev.weapon, // pass by ref
                ev.ability_multiplier_bp,
                roll
            );

            // emit health changed event (keep this cheap)
            health_changed.write(HealthChanged {
                entity: ev.target,
                health: health.current,
                max_health: health.max,
            });

            // todo: add lightweight hooks for crits / VFX in other systems that listen to HealthChanged
            let _ = was_crit; // placeholder to avoid unused warnings if you don't use it yet
        }
    }
}

/// Simple label (or event) to mark death handling systems
#[derive(SystemSet, Clone, Hash, PartialEq, Eq, Debug)]
pub enum CombatSet {
    Damage,
    Death,
}

/// System: mark / process dead entities
pub fn death_system(mut commands: Commands, query: Query<(Entity, &Health), Changed<Health>>) {
    for (entity, health) in query.iter() {
        if health.current <= 0 {
            // handle death: state change, drop loot, remove components, or despawn
            // keep heavy effects (spawn particles, sound) in a separate system that listens to DeathEvent.
            commands.entity(entity).remove::<(/* combat components we want to clear */)>();
            // And maybe we'll despawn them.
            // commands.entity(entity).despawn_recursive();
        }
    }
}

File: python/src/game/systems/combat.py
"""
Combat systems for the ECS game logic.

Converted from Rust to Python 3.14
"""

from dataclasses import dataclass
from enum import Enum
from typing import Protocol, Any, Optional

from ..components.combat import (
    AttackType,
    Health,
    HealthChanged,
    Stats,
    resolve_damage,
)
from .rng import GameRng


class Entity(Protocol):
    """Protocol for entity types"""
    id: Any


class EventReader(Protocol):
    """Protocol for event readers"""
    def read(self):
        """Read events from the event queue"""
        ...


class EventWriter(Protocol):
    """Protocol for event writers"""
    def write(self, event: Any) -> None:
        """Write an event to the event queue"""
        ...


class Query(Protocol):
    """Protocol for ECS queries"""
    def get(self, entity: Entity):
        """Get component for entity"""
        ...

    def get_mut(self, entity: Entity):
        """Get mutable component for entity"""
        ...

    def iter(self):
        """Iterate over query results"""
        ...


class Commands(Protocol):
    """Protocol for ECS commands"""
    def entity(self, entity: Entity):
        """Get entity commands"""
        ...


@dataclass
class AttackEvent:
    """
    Event describing a single attack intent. Minimal and cheap.

    Attributes:
        attacker: Entity performing the attack
        target: Entity being attacked
        weapon: Type of attack being used
        ability_multiplier_bp: Optional ability multiplier in basis points (10000 = 1.0)
    """
    attacker: Any  # Entity
    target: Any  # Entity
    weapon: AttackType
    ability_multiplier_bp: Optional[int] = None


def apply_attack_events(
    ev_reader: EventReader,
    health_q: Query,
    stats_q: Query,
    rng: GameRng,
    health_changed: EventWriter,
) -> None:
    """
    Consume AttackEvent and apply damage.

    Note: `events` must be provided and the event type registered with the app.

    Args:
        ev_reader: Event reader for AttackEvent
        health_q: Query for Health components (mutable)
        stats_q: Query for Stats components
        rng: Game RNG resource
        health_changed: Event writer for HealthChanged events
    """
    for ev in ev_reader.read():
        # Get attacker stats
        try:
            attacker_stats = stats_q.get(ev.attacker)
        except Exception:
            # Attacker missing stats -> skip
            continue

        # Get target stats
        try:
            target_stats = stats_q.get(ev.target)
        except Exception:
            # Target missing stats -> skip
            continue

        # Apply damage if target has health
        try:
            health = health_q.get_mut(ev.target)
            roll = rng.next_u16_in_basis()
            damage, was_crit = resolve_damage(
                health,
                attacker_stats,
                target_stats,
                ev.weapon,
                ev.ability_multiplier_bp,
                roll
            )

            # Emit health changed event (keep this cheap)
            health_changed.write(HealthChanged(
                entity=ev.target,
                health=health.current,
                max_health=health.max,
            ))

            # TODO: add lightweight hooks for crits / VFX in other systems
            # that listen to HealthChanged
            _ = was_crit  # Placeholder to avoid unused warnings

        except Exception:
            # Target doesn't have health component
            continue


class CombatSet(Enum):
    """Simple label for combat system ordering"""
    DAMAGE = "damage"
    DEATH = "death"


def death_system(commands: Commands, query: Query) -> None:
    """
    System to mark/process dead entities.

    Args:
        commands: Commands for entity manipulation
        query: Query for entities with Health component that have changed
    """
    for entity, health in query.iter():
        if health.current <= 0:
            # Handle death: state change, drop loot, remove components, or despawn
            # Keep heavy effects (spawn particles, sound) in a separate system
            # that listens to DeathEvent.

            # Remove combat components we want to clear
            # commands.entity(entity).remove(...)

            # Maybe despawn the entity
            # commands.entity(entity).despawn_recursive()
            pass
